File: Data Structures & Algorithms/path-sum/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        
        res = []

        def dfs(root, curSum):
        
            # Base Case
            if not root:
                return

            curSum += root.val
            if not root.left and not root.right:
                res.append(curSum)
                return
            
            # Recursive Case
            dfs(root.left, curSum)
            dfs(root.right, curSum)

        dfs(root, 0)

        return targetSum in res

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-0.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

    private int maxSum = Integer.MIN_VALUE;

    public int maxPathSum(TreeNode root) {
        dfs(root);
        return maxSum;
    }

    private int dfs(TreeNode node) {
        
        if (node == null) return 0;

        int leftMax  = Math.max(0, dfs(node.left));
        int rightMax = Math.max(0, dfs(node.right));

        int localMax = leftMax + rightMax + node.val;
        maxSum = Math.max(maxSum, localMax);

        return Math.max(leftMax, rightMax) + node.val;
    }
}
